### PR TITLE
Interstellar - Wallet Phase 2 (follow-up)

### PR DESCRIPTION
**Pull request by nashjl**
#### Project Abstract
Interstellar is a new universal layer to access Web3.
 Wallet Phase 1: [w3f/Grants-Program#1354](https://github.com/w3f/Grants-Program/pull/1354)

#### Grant [level](https://github.com/w3f/Grants-Program#level_slider-levels)

- [ ] **Level 1**:  Up to $10,000, 2 approvals
- [x] **Level 2**:  Up to $30,000, 3 approvals
- [ ] **Level 3**:  Unlimited, 5 approvals (for >$100k: Web3 Foundation Council approval)

#### Application Checklist

- [x] The [application template](https://github.com/w3f/Grants-Program/blob/master/applications/application-template.md) has been copied and aptly renamed (`project_name.md`).
- [x] I have read the [application guidelines](https://github.com/w3f/Grants-Program/blob/master/docs/Support%20Docs/grant_guidelines_per_category.md).
- [ ] A BTC, Ethereum (USDC/DAI) or Polkadot/Kusama (aUSD/USDT) address for the payment of the milestones is provided in the application.
- [x] The software delivered for this grant will be released under an open-source license specified in the application.
- [x] The initial PR contains only one commit (squash and force-push if needed).
- [x] The grant will only be announced once the first milestone [has been accepted](https://github.com/w3f/Grant-Milestone-Delivery#process) (see the [announcement guidelines](https://github.com/w3f/Grants-Program/blob/master/docs/Support%20Docs/announcement-guidelines.md)).
- [ ] I prefer the discussion of this application to take place in a private Element/Matrix channel. My username is: `@_______:matrix.org` (change the homeserver if you use a different one)

### CODE REVIEW

**Review by Noc2 — CHANGES_REQUESTED** [review]
Thanks a lot for the follow-up application. I have one initial question: What are your long-term goals with this project? Did you already try to apply for treasury funding with this? If you plan to help different parachains and, if I understand you correctly, you're already talking to parachain teams, it might be the logical next step. Also, regarding the long-term funding of this project. If your long-term plans are instead to release your own parachain, then the substrate builders program might be better suited for you, see https://substrate.io/ecosystem/substrate-builders-program/

**Review by semuelle — CHANGES_REQUESTED** [review]
Thanks for the follow-up application, @nashjl! I find it generally quite interesting, I just have trouble understanding the overall architecture of Interstellar and the relationship between its components. Perhaps you could add an architecture or sequence diagram containing Integritee, Interstellar and Polkadot and what data is exchanged and how?

File: applications/Interstellar-network2.md [hunk 1]
```diff
@@ -0,0 +1,304 @@
...
+| **0c.** | Testing and Testing Guide    | Core functions will be fully covered by comprehensive unit tests to ensure functionality and robustness. In the guide, we will describe how to run these tests.                                                              |
+| **0d.** | Docker                       | We will provide a Dockerfile(s) that can be used to test all the functionality delivered with this milestone.                                                                                                                |  |  |  |
+|      1. | Garble Circuit pallet update | We will optimize the garbled circuit evaluation scheme to get better performance                                                                                                                                             |
+|      2. | Display Circuit update       | We will modify the current display circuit to enable a smoother user experience more comfortable for the user when circuit is evaluated                                                                                      |
```
[the remark this answers is not in the document]
Comment (nathanprat) [reply applications/Interstellar-network2.md:246]: That is not the easiest one to evaluate... But it was you who evaluated the latest milestone, right?
If yes, you may have noticed that the app was running at maybe ~1 FPS and that was pretty much unreadable.
The goal of this milestone is to target at least 60FPS, and ideally 120+.
At those frame rates, it is a way more pleasant user experience.

Practically, we will have to refactor the garbling lib we are using(`swanky`) and add various optimizations from a white paper.

We will also probably add a "dynamic" feature(ie fading) to make the screen more readable, sacrificing a bit of security.
That should be user-configurable and/or server-configurable; eg based on the amount.

File: applications/Interstellar-network2.md [hunk 2]
```diff
@@ -0,0 +1,304 @@
...
+|      1. | Substrate Module - social recovery | We will integrate social recovery pallet with our chain                                                                                                                                                                      |
+|     2.a | Android Wallet                     | We will update android app to handle social recovery                                                                                                                                                                         |
+|     2.b | iOS Wallet                         | We will update iOS app to handle social recovery                                                                                                                                                                             |
+|      3. | Substrate Integritee chain         | POC to demonstrate instant onboarding and easy airdrop                                                                                                                                                                       |
```
[the remark this answers is not in the document]
Comment (nathanprat) [reply applications/Interstellar-network2.md:229]: This milestone would be only Android and iOs apps. Ideally deployed through Firebase and TestFlight.
With added features of social recovery, instant onboarding/airdrop, and MVP UX(ie no app closing at the end of the tx, proper error handling, etc).

The goal is to deliver MVP-grade apps, minus the garbling optimizations of the next milestone.

**Comment by nathanprat** [comment]
Hello,
@semuelle cf the comments for details

For an overview of our architecture, you can find info [here](https://book.interstellar.gg/GCF.html) and [there](https://book.interstellar.gg/TTVP.html).
It is a bit dense, and probably not easily "discoverable". We are open to feedback!

**Comment by nashjl** [comment]
@Noc2 Thanks a lot for your question.

Our long-term goal is to launch our own parachain and help current and future Polkadot parachains. 

As a small team with limited resources, we recognize that we may be too early stage to apply for the Substrate Builders Program or treasury funding. 
Instead, our short-term goal is to develop the missing key features of our solution, such as an iOS version (as suggested by the grant team), and Polkadot and Ethereum transaction capabilities.

Once we've developed these features, we can showcase a more complete and attractive version of our solution to both potential investors and parachain teams we are in conversations with. 
This step will help us secure a pre-seed funding round, which will not only allow us to continue developing our solution, but also expand our team and position us for future opportunities like the Substrate Builders Program and treasury funding.

We believe that by following this roadmap, we can build a compelling solution that benefits the Polkadot ecosystem and ultimately helps us achieve our long-term goal of launching our own parachain.

**Comment by semuelle** [comment]
Thanks for the replies, @nathanprat & @nashjl. I would like to see this grant going forward, but there are still some details I cannot quite figure out. The book you linked helped a little, but maybe I'm too dense. For example, where/how does the "Integritee pallet" (D1.3) connect to Ethereum? Going through the diagram in the project details, what are the striped and dotted components? Mobile app, Interstellar chain and Integritee? Regarding M5, I would suggest you either specify more clearly what these optimisations will look like, or you remove the milestone as a whole and include that in a future proposal.

**Comment by nashjl** [comment]
Hi @semuelle, Thanks for your message.  Now that you mention it, it’s true that the diagram is confusing. We updated it here   Architecture overview
 Let us know if that gives you a better overview.

>Where/how does the "Integritee pallet" (D1.3) connect to Ethereum?

>"Integritee pallet - Ethereum signer"

 is only the signer.

For the architecture we have two options:
- we have no client at all app-side; and our stack publishes the transaction to the appropriate network cf M1.4 "Integritee pallet - submit transaction"
- our stack only handles the wallet(i.e. the signatures) and then returns a raw (signed) TX to the mobile which can be send to the appropriate network with a local (light?) client

I would argue that from the point of view of this milestone this is an implementation detail:
- the complexity is handling the keys and signing the TX; publishing a signed TX is a solved issue: it could be done with an already existing pallet(?), or via a basic RPC and a light client we could run as part of our stack, or even via a 3rd party API(at least for now for demo/MVP purposes).

>"Regarding M5, I would suggest you either specify more clearly what these optimisations will look like, or you remove the milestone as a whole and include that in a future proposal."

I don't think writing about implementation details up-front is a good idea: these are almost guaranteed to become irrelevant after the implementation is done.
Would you be OK with:
- specifying a performance target(e.g. 60 fps)
+ optionally the link to the white paper?

The goal of M5 is to have a functional MVP. Without these optimizations, we won’t be able to deliver a proper MVP demo (we are currently at <1 FPS)

**Review by semuelle — COMMENTED** [review]
Thanks for the update. Sounds good to me.

> Would you be OK with:
> 
>     * specifying a performance target (e.g., at least 60 fps ~ previous grant M3 performance)?
>     * optionally the link to the white paper related to a permutation based garbling scheme designed for performance.

Either one is fine. I just want to make sure that, when this is delivered, we have a way of actually evaluating it. Any details you can add about what will/might/could be done would help.

**Comment by nashjl** [comment]
@semuelle Thanks for your response, we have updated M5 with what you agreed on. We will also add fps in the logs.
We can assure you that regarding the evaluation of M5, you will see a major improvement!

**Review by semuelle — APPROVED** [review]
Thanks for the update.

**Review by Noc2 — CHANGES_REQUESTED** [review]
Sorry for the late reply, and thanks for the detailed response. Something to be aware of: the SBP also supports projects financially, as far as I know, and helps you get in touch with VCs. Apart from this, I still recommend reducing the scope of this application or applying for treasury funding. I will share the application again with the rest of the team for additional feedback.

**Comment by nashjl** [comment]
Thanks for getting back @Noc2! 

We are making progress with our partnerships and would really like to continue working with the Web3 Foundation. Following your recommendation, we could reduce the scope of this application to 2-3 milestones. Regarding alternative funding, we will check again with the SBP but are also currently looking at the Ethereum Support Program which could help with developing the MVP before applying for treasury funding. 

Additionally, we just applied to talk at Polkadot Decoded 2023 which might be a good possibility to get more traction (if accepted). 

Please let us know if this approach makes sense, so we can propose an updated scope and sync with ESP.

**Comment by keeganquigley** [comment]
Thanks for the reply @nashjl it's possible that if you reduce the scope/cost of the application it might have a better chance of getting approved. Perhaps you could focus purely on the PoC of the garbled circuits and pallets and less on the app deployment; if you lowered it to a level 2 grant you would only need 3 approvals.

**Comment by nashjl** [comment]
Thanks for the recommendation @keeganquigley! 

We modified the PR into a level 2 application, focusing purely on the garbled circuits and pallets. 👍

**Review by semuelle — APPROVED** [review]

File: applications/Interstellar-network2.md [hunk 3]
```diff
@@ -0,0 +1,201 @@
+# Interstellar - Wallet Phase 2
+
+
+- **Team Name:** Interstellar
+- **Payment Address:** TBD
+- **[Level](https://github.com/w3f/Grants-Program/tree/master#level_slider-levels):** 3
```
Comment (keeganquigley) [inline applications/Interstellar-network2.md:6]: Thanks for changing the scope! Can you update the application to reflect a level 2 grant? Also, can you update the payment address now or do you plan to submit an amendment at a later time?

Comment (keeganquigley) [reply applications/Interstellar-network2.md:6]: And can you change the checked box on the PR as well? Thanks!

**Review by keeganquigley — APPROVED** [review]
Thanks @nashjl I believe a contract will need to be signed in order to get paid in fiat, but I will double check with the team tomorrow. Other than that I'm happy to support it.

**Comment by nashjl** [comment]
Great. Thanks a lot @keeganquigley for your support!

**Review by Noc2 — APPROVED** [review]
I'm happy to go ahead with it. For fiat payment, we don't need to sign a separate contract. See https://github.com/w3f/Grants-Program/blob/master/applications/application-template.md 

> "In the case of fiat payment, please share your bank account privately with grants@web3.foundation via your contact email (see below) and enter the date when you shared the information with us (e.g. Fiat 24.12.1971, 11:59) here."

**Review by dsm-w3f — APPROVED** [review]
Thank you for your application. Reading the discussion, I think is a good strategy to continue working and trying to find other sources of funding. I'm happy to support the project, it seems to be a very promising one. One idea is to look for mobile phone companies (hardware ones) that could embed the technology in their products. However, I strongly advise trying to get a patent before doing that. Good luck in finding more supporters.

**Review by keeganquigley — APPROVED** [review]

**Comment by nashjl** [comment]
Thanks a lot for your support @Noc2, @semuelle, @keeganquigley and @dsm-w3f!

Thank you also for your feedback, guidance, advice and ideas. Means a lot!

**Pull request merged**